modules/mauth-authenticator-http4s/src/main/scala/com/mdsol/mauth/http4s/MauthPublicKeyProvider.scala

```diff
@@ -34,12 +35,15 @@ class MauthPublicKeyProvider[F[_]: Concurrent: Logger](configuration: Authentica
     * @return { @link PublicKey} registered in MAuth for the application with given appUUID.
     */
   override def getPublicKey(appUUID: UUID): F[Option[PublicKey]] = memoizeF(Some(configuration.getTimeToLive.seconds)) {
-    val uri = new URI(configuration.getBaseUrl + getRequestUrlPath(appUUID))
-    val signedRequest = signer.signRequest(UnsignedRequest.noBody("GET", uri, headers = Map.empty))
-    signedRequest
-      .toHttp4sRequest[F]
-      .flatMap(req => client.run(req).use(retrievePublicKey))
-  }
+    Async[F].defer {
+      val uri = new URI(configuration.getBaseUrl + getRequestUrlPath(appUUID))
+      val signedRequest = signer.signRequest(UnsignedRequest.noBody("GET", uri, headers = Map.empty))
+      signedRequest
+        .toHttp4sRequest[F]
+        .flatMap(req => client.run(req).use(retrievePublicKey))
+    }.memoize
+  }.flatten
```

Comment:
If an error happened we should remove the IO from the cache.

Reply:
Yup, fair enough. Done that now.